### PR TITLE
Test refactor

### PR DESCRIPTION
**Pull request by halkony**
Refactors the current test into more readable patterns.

I left some comments about improving syntax for testing. Additionally, breaking out one of the tests broke another part of the test and you'd probably have an easier time debugging that since you knows gear trains better.

### CODE REVIEW

File: test_simulation.py [hunk 1]
```diff
@@ -4,55 +4,77 @@
...
-    gtrain.add_gear(g0)
-    gtrain.add_gear(g1)
-    gtrain.add_gear(g2)
+    gtrain.add_gear(gear1)
+    gtrain.add_gear(gear2)
+    gtrain.add_gear(gear3) # NDH: i think this reveals we should add a *args to the aggregate constructors or otherwise abstract GearTrains representation
+    return gtrain
+
+def test_geartrain_implementation_in_simnode(gear1, gear2, gear3, gear_train):
+    # NDH: I had to keep these two lines up here or this test would fail for some reason
```
Comment (halkony) [inline test_simulation.py:37]: Wasn't able to debug this easily.

File: test_simulation.py [hunk 2]
```diff
@@ -4,55 +4,77 @@
...
-    ## notice direction switched even though g1 did not rotate, since direction is an input variable rather than a state variable
+    gear2_state_before = gear2.get_state_variables()
+    gear2_input_before = gear2.get_input_variables()
+    gear3_state_before = gear3.get_state_variables()
+    gear3_input_before = gear3.get_input_variables()
+
+    assert gear_train.update()
+    # NDH: I see in the implementation update this returns False if unsuccessful.
+    # It's more pythonic to define an exception "UpdateException" that can be thrown from those classes.
+    # That way we can catch it where we need to without doing boolean compares. This is especially useful when we start working with try catch game loop.
```
Comment (halkony) [inline test_simulation.py:61]: ^

File: test_simulation.py [hunk 3]
```diff
@@ -4,55 +4,77 @@
...
+# arrange, act, assert
 
-def test_geartrain_implementation_in_simnode():
-    g0 = Gear(10)
-    g1 = Gear(20)
-    g2 = Gear(10)
+@pytest.fixture
+def gear1():
+    gear = Gear(10)
+    gear.probability_of_breaking_per_tooth_click = 0 # NDH: having to move this into declaration feels like anti pattern
```
Comment (halkony) [inline test_simulation.py:13]: Simnodes should probably be deterministic on declaration. If we end up with like 500,000+ nodes at runtime, updating a bunch of properties for determinism wouldnt scale too well within the application.

File: test_simulation.py (hunk 2 above)
Comment (diongbridger) [reply test_simulation.py:61]: I think there is a rationale for both a boolean return value and in some circumstances an exception. The return value being false doesn't necessarily indicate a logic error, just that for the SimGraph in its current state with the given inputs, no applicable changes could be computed (think of trying to turn a mesh of three gears locked in a triangle). We could throw an exception if we are sure that an update attempt should result in some changes but fails.

File: test_simulation.py (hunk 1 above)
Comment (diongbridger) [reply test_simulation.py:37]: I'll try to reproduce on my end.

File: test_simulation.py (hunk 3 above)
Comment (diongbridger) [reply test_simulation.py:13]: Agreed. Although there are some circumstances in which an internal or external influence could alter a property of the node. E.g. wear & tear on machine parts. Setter methods maybe?

**Review by diongbridger — APPROVED** [review]
Looks good!

File: test_simulation.py (hunk 2 above)
Comment (halkony) [reply test_simulation.py:61]: So are you saying that, as far as update() is concerned, the possible results would be

1. Returns True, no logic error and node is updated
2. Returns False, no logic error in implementation and node is reset (rather, it never changed)
3. Returns False, logic error in implementation then node is reset (this is probably where we should throw exception)

Is this a broad interface decision or just for SimGraph?
I see value in knowing if the state of a node changed after an update. I tend to separate function with side effects from my functions that return values. But considering the problem space, it's probably a good interface decision. Storing old objects in memory for comparison would be annoying.

**Pull request merged**